src/lib/parameters/autosave.cpp

```diff
@@ -127,5 +162,15 @@ void ParamAutosave::Run()
 	} else {
 		_retry_count = 0;
 	}
+
+	{
+		AtomicTransaction transaction;
```

Comment:
This doesn't really work correctly; I believe you should always save the parameters once *after* the shutdown callback.

Now, if the shutdown callback occurs in the middle of Run method, it will set the ready_to_shutdown and call complete below before it has executed the exact work item scheduled in the shutdown callback.

So the potential error would occur if 1) Run method is executing, in the middle of saving parameters 2) some parameter is changed (which is not saved yet in this Run method) 3)  shutdown callback occurs before Run method reaches the "if" below (note that AtomicTransaction scope ends before the If).

IMHO, you can remove the AtomicTransaction here and in callback function, but read the "ready_to_shutdown" parameter above before calling param_save_default ?